

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Custom pipelines &mdash; Katuali 0.2.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Medaka training pipeline" href="medaka_train.html" />
    <link rel="prev" title="Basic Usage and Tests" href="tests.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Katuali
          

          
          </a>

          
            
            
              <div class="version">
                0.2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Basic Usage and Tests</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Custom pipelines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basecalling">Basecalling</a></li>
<li class="toctree-l2"><a class="reference internal" href="#assembly">Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="#polishing">Polishing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipeline-restrictions">Pipeline restrictions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#starting-from-existing-basecalls">Starting from existing basecalls</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-read-coverage-depth">Calculating read coverage depth</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-subsampled-datasets">Creating subsampled datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#subsampling-a-single-reference-contig">Subsampling a single reference contig</a></li>
<li class="toctree-l2"><a class="reference internal" href="#subsampling-specified-regions">Subsampling specified regions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#medaka-training-pipeline">Medaka training pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="medaka_train.html">Medaka training pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Pipeline configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Katuali</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Custom pipelines</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="custom-pipelines">
<span id="introduction"></span><h1>Custom pipelines<a class="headerlink" href="#custom-pipelines" title="Permalink to this headline">¶</a></h1>
<p>One of the main aims of <cite>Katuali</cite> is to enable the bolting-together of any
number of basecallers, assemblers, and polishers into flexible (i.e. not
predefined) pipelines. Basecallers, assemblers, and polishers can be
interchanged because their Snakefile rules have standardised inputs and
outputs.</p>
<p><cite>Katuali</cite> makes strong use of <cite>Snakemake</cite>’s powerful <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#wildcards">wildcards</a>
to extract pipeline parameters from the target file path. For example
which basecaller or assembler to use as well as parameters for these tools.</p>
<p>For example, while</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/guppy/basecalls.fasta
</pre></div>
</div>
<p>will basecall with guppy.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/scrappie/basecalls.fasta
</pre></div>
</div>
<p>will perform basecalling with scrappie (starting from <cite>./run1/reads</cite>),</p>
<p>Each step of a multi-step pipeline stores its outputs in a subdirectory of the
previous stage. In this way, we can define a multistep-pipeline to basecall
with guppy, assemble with canu, then polish with medaka, and finally run
nanopolish by contructing the target:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/guppy/canu_gsz_4.0M/medaka/nanopolish/consensus.fasta
</pre></div>
</div>
<p>This nested working directory stores the data in such a way that is it obvious
what went into and out of each stage of the pipeline.</p>
<p>It also enables forks in the pipelines which might share basecalls (or other
intermediate results), but differ in assembly or consensus methods.</p>
<p>For example, if we wish to basecall with guppy, assemble with canu, run
medaka on the canu assembly, and seperately run nanopolish on the canu assembly,
we could use the targets:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/guppy/canu_gsz_4.0M/nanopolish/consensus.fasta run1/basecall/guppy/canu_gsz_4.0M/medaka/consensus.fasta
</pre></div>
</div>
<p><cite>Snakemake</cite> will create a graph of tasks to perform the common basecall
and assembly tasks, then run separately nanopolish and medaka from the same
inputs (and in parallel, given enough resource).</p>
<div class="section" id="basecalling">
<h2>Basecalling<a class="headerlink" href="#basecalling" title="Permalink to this headline">¶</a></h2>
<p>Supported basecallers include guppy, scrappie, flappie:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/guppy/basecalls.fasta
katuali run1/basecall/scrappie/basecalls.fasta
katuali run1/basecall/flappie/basecalls.fasta
</pre></div>
</div>
</div>
<div class="section" id="assembly">
<h2>Assembly<a class="headerlink" href="#assembly" title="Permalink to this headline">¶</a></h2>
<p>Reads can be assembled in two ways at present:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># assemble with canu, specifying the genome size in the target name.</span>
katuali run1/basecall/scrappie/canu_gsz_4.0M/consensus.fasta

<span class="c1"># use pomoxis mini_assemble to assemble with miniasm, then form consensus</span>
<span class="c1"># with racon</span>
katuali run1/basecall/scrappie/miniasm_racon/consensus.fasta
</pre></div>
</div>
</div>
<div class="section" id="polishing">
<h2>Polishing<a class="headerlink" href="#polishing" title="Permalink to this headline">¶</a></h2>
<p>Sequences can be polished with Racon, Nanopolish and medaka to create higher
accuracy consensus sequences. Consensus methods can also be combined (e.g.
racon/medaka) meaning that the input to medaka will be the racon consensus.
The last example requests two rounds of medaka.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/guppy_flipflop/canu_gsz_4.0M/racon/nanopolish/consensus.fasta
katuali run1/basecall/guppy_flipflop/canu_gsz_4.0M/racon/medaka_flipflop/consensus.fasta
katuali run1/basecall/guppy_flipflop/canu_gsz_4.0M/racon/medaka_flipflop/medaka_flipflop/consensus.fasta
</pre></div>
</div>
</div>
<div class="section" id="pipeline-restrictions">
<h2>Pipeline restrictions<a class="headerlink" href="#pipeline-restrictions" title="Permalink to this headline">¶</a></h2>
<p><cite>Katuali</cite> aims to be as flexible as possible, but there are some obvious
restrictions:</p>
<blockquote>
<div><ul class="simple">
<li><p>basecalling must be performed before assembly.</p></li>
<li><p>assembly must come before polishing (use of polishing targets to
error correct reads is not supported).</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="starting-from-existing-basecalls">
<span id="starting-from-basecalls"></span><h2>Starting from existing basecalls<a class="headerlink" href="#starting-from-existing-basecalls" title="Permalink to this headline">¶</a></h2>
<p>If you have already basecalled your data, mocking out the working space as if
katuali had basecalled allows any derived targets to be created.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Input files</span>
<span class="nv">BASECALLS</span><span class="o">=</span>/path/to/basecalls.fastq
<span class="nv">SUMMARY</span><span class="o">=</span>/path/to/sequencing_summary.txt

<span class="c1"># These should be set as in the config.yaml file used for running the</span>
workflow. RUN is <span class="c1"># the top level key of the DATA section</span>
<span class="nv">RUN</span><span class="o">=</span>run1
<span class="nv">BASECALLER</span><span class="o">=</span>guppy_flipflop
<span class="nv">IN_POMOXIS</span><span class="o">=</span>~/git/pomoxis/venv/bin/activate

<span class="c1"># ...no need to edit below here</span>
<span class="nv">BCDIR</span><span class="o">=</span><span class="si">${</span><span class="nv">RUN</span><span class="si">}</span>/basecall/<span class="si">${</span><span class="nv">BASECALLER</span><span class="si">}</span>/
mkdir -p <span class="si">${</span><span class="nv">BCDIR</span><span class="si">}</span>
mkdir <span class="si">${</span><span class="nv">RUN</span><span class="si">}</span>/reads
ln -s <span class="si">${</span><span class="nv">SUMMARY</span><span class="si">}</span> <span class="si">${</span><span class="nv">BCDIR</span><span class="si">}</span>/sequencing_summary.txt

<span class="nb">source</span> <span class="si">${</span><span class="nv">IN_POMOXIS</span><span class="si">}</span>
seqkit fq2fa <span class="si">${</span><span class="nv">BASECALLS</span><span class="si">}</span> &gt; <span class="si">${</span><span class="nv">BCDIR</span><span class="si">}</span>/basecalls.fasta
</pre></div>
</div>
<p>Now katuali can be run as normal, for example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali --configfile my_config.yaml standard_assm_polish
</pre></div>
</div>
</div>
<div class="section" id="calculating-read-coverage-depth">
<h2>Calculating read coverage depth<a class="headerlink" href="#calculating-read-coverage-depth" title="Permalink to this headline">¶</a></h2>
<p>It is often useful to know the read coverage depth of a dataset.
This requires a reference.fasta to be specified in the config to which the reads will be aligned.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DATA</span><span class="p">:</span>
    <span class="s">&#39;run1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="l l-Scalar l-Scalar-Plain">:/path/to/ref.fasta</span>
</pre></div>
</div>
<p>The read coverage depth can then be calculated as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/scrappie/align/depth
</pre></div>
</div>
<p>The depth directory will contain a text file per reference contig with coverage
vs genomic coordinate, as well as a file containing summary statistics for all
contigs.</p>
</div>
<div class="section" id="creating-subsampled-datasets">
<h2>Creating subsampled datasets<a class="headerlink" href="#creating-subsampled-datasets" title="Permalink to this headline">¶</a></h2>
<p>Katuali also supports the generation of datasets with even coverage over a
reference at a given depth.
This requires a reference.fasta to be specified in the config to which the reads will be aligned.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DATA</span><span class="p">:</span>
    <span class="s">&#39;run1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="l l-Scalar l-Scalar-Plain">:/path/to/ref.fasta</span>
</pre></div>
</div>
<p>Once the reference is the config, running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/scrappie/align/all_contigs/25X/miniasm_racon/consensus.fasta
</pre></div>
</div>
<p>will perform the following steps:</p>
<blockquote>
<div><ul class="simple">
<li><p>basecall the reads to create:
<cite>run1/basecall/scrappie/basecalls.fasta</cite></p></li>
<li><p>align the basecalls to the reference to create:
<cite>run1/basecall/scrappie/align/calls2ref.bam</cite></p></li>
<li><p>subsample all contigs in the .bam file to 25X to create (in one step):
<cite>run1/basecall/scrappie/align/all_contigs/25X/basecalls.fasta</cite></p></li>
<li><p>perform a ref-guided assembly and racon consensus to create:
<cite>run1/basecall/scrappie/align/all_contigs/25X/miniasm_racon/consensus.fasta</cite></p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The rule to create subsampled datasets differs from other rules in
that it creates two levels of nested directories in a single step (in this case
<cite>all_contigs/25X</cite>). The extraction of specific regions/contigs without
subsampling to a specific depth is not currently supported.</p>
</div>
</div>
<div class="section" id="subsampling-a-single-reference-contig">
<h2>Subsampling a single reference contig<a class="headerlink" href="#subsampling-a-single-reference-contig" title="Permalink to this headline">¶</a></h2>
<p>It is also possible to subsample just one of the contigs in your reference by
specifying targets such as:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali run1/basecall/scrappie/align/ecoli_SCS110_plasmid2/25X/miniasm_racon/consensus.fasta
</pre></div>
</div>
<p>which will just process the reference sequence <cite>ecoli_SCS110_plasmid2</cite>.</p>
</div>
<div class="section" id="subsampling-specified-regions">
<h2>Subsampling specified regions<a class="headerlink" href="#subsampling-specified-regions" title="Permalink to this headline">¶</a></h2>
<p>It is also possible to subsample only specified regions specifed as samtools
strings:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">REGIONS</span><span class="o">=</span><span class="s2">&quot;ecoli_SCS110_chromosome:50000-150000 ecoli_SCS110_chromosome:200000-250000&quot;</span>
katuali run1/basecall/scrappie/align/my_regions/25X/miniasm_racon/consensus.fasta --config <span class="nv">REGIONS</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$REGIONS</span><span class="s2">&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="medaka-training-pipeline">
<span id="train-medaka"></span><h2>Medaka training pipeline<a class="headerlink" href="#medaka-training-pipeline" title="Permalink to this headline">¶</a></h2>
<p>It is possible to train medaka models starting from folders of fast5s in a
single command once the config has been modified to reflect your input data
(fast5s and genomes for each run as well as training and evaluation region
definitions).</p>
<p><cite>MEDAKA_TRAIN_REGIONS</cite> and <cite>MEDAKA_EVAL_REGIONS</cite> define regions for training
and evaluation.  In the example below we train from the <cite>minion</cite> run using
<cite>ecoli</cite> and <cite>yeast</cite> contigs in the reference and evaluate on the <cite>gridion</cite> run
using the contigs <cite>ecoli</cite>, <cite>yeast</cite> and <cite>na12878_chr21</cite> in the reference.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DATA</span><span class="p">:</span>
    <span class="s">&#39;MinIonRun1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
    <span class="s">&#39;MinIonRun2&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
    <span class="s">&#39;GridIonRun1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;na12878_chr21&#39;</span><span class="p p-Indicator">]</span>
    <span class="s">&#39;GridIonRun2&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;na12878_chr21&#39;</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
<p>Running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali all_medaka_train_features --keep-going
</pre></div>
</div>
<p>will:</p>
<ul class="simple">
<li><p>basecall all the runs</p></li>
<li><p>align each run to its reference</p></li>
<li><p>create subsampled sets of basecalls over the desired regions and depths</p></li>
<li><p>assemble those sets of basecalls</p></li>
<li><p>create medaka training features for all those sets</p></li>
</ul>
<p>Running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali medaka_train_replicates --keep-going
</pre></div>
</div>
<p>will do all the tasks of <code class="docutils literal notranslate"><span class="pre">all_medaka_train_features</span></code> and additionally launch
multiple medaka model-training replicates.</p>
<p>If some of your input runs have insufficient coverage-depth for some of the
training regions, some of the training feature files will not be made. In this
case the config flag <code class="docutils literal notranslate"><span class="pre">USE_ONLY_EXISTING_MEDAKA_FEAT</span></code> can be set to true to allow
katuali to train using only those features which exist already:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">USE_ONLY_EXISTING_MEDAKA_FEAT</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">true</span>
</pre></div>
</div>
<p>Refer to comments in the config (katuali/data/config.yaml) to see how this process
can be controlled.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="medaka_train.html" class="btn btn-neutral float-right" title="Medaka training pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tests.html" class="btn btn-neutral float-left" title="Basic Usage and Tests" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018-19, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>